### PR TITLE
[WIP] [DSCP-67] test interpreter for query dsl

### PR DESCRIPTION
**Pull request by lambda76**
- [x] Simple DSL test interpreter in pure code

- [x] Write test queries running in SimpleTxDB

### CODE REVIEW

File: src/Disciplina/DB/DSL/SimpleTxDB.hs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+instance MonadSearchTxObj SimpleTxDB where
+  runTxQuery = runSimpleTxQuery
+  runTxsQuery = runSimpleTxsQuery
+  runObjQuery = undefined -- TODO, implement this one, but how are Object defined?
+
+instance Serialise Address where
+  encode (Address h) = encode h
+  decode = Address <$> decode
+
+-- | TODO, where should we put these?
```
Comment (flyingleafe) [inline src/Disciplina/DB/DSL/SimpleTxDB.hs]: Separate 'Serialise' modules, I think. Soon there will be my PR on DSCP-61, you'll see how I see it.

File: src/Disciplina/DB/DSL/SimpleTxDB.hs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+-- | Find Tx in db with hash == h
+runSimpleTxQuery :: QueryTx -> SimpleTxDB (Maybe PrivateTx)
+runSimpleTxQuery (SELECTTx _ (TxIdEq (h :: PrivateTxId))) =
+  safeHead.filter ((==h).hash) <$> ask
+
+-- | Find Txs in db with SubjectId == a
+runSimpleTxsQuery :: QueryTxs -> SimpleTxDB [PrivateTx]
+runSimpleTxsQuery (SELECTTxs _ (TxSubjectIdEq sId)) =
+  filter ((==sId).ciSubject._ptxCourseId) <$> ask
+
```
Comment (flyingleafe) [inline src/Disciplina/DB/DSL/SimpleTxDB.hs]: I think that it's a bit confusing when you leave a blank line between pattern matches of the same function.

File: src/Disciplina/DB/DSL/SimpleTxDB.hs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+testFindTx2 :: Maybe PrivateTx
+testFindTx2 = runFindTx $ mkPrivateTx 'a' 'k'
+
+
+-- | run query
+-- | I want to be able to get { [<tx>] | P(<ordered_tx_param>) }, where P is a system of (in)equalities
+runFindTxs :: QueryTxs -> [PrivateTx]
+runFindTxs q = runReader (getSimpleTxDB . runQuery $ q) simpleTxDB
+
+-- | TODO, move tests to test module
```
Comment (flyingleafe) [inline src/Disciplina/DB/DSL/SimpleTxDB.hs]: :heavy_plus_sign: . Everything after `runSimpleTxsQuery` should be in test module.

File: src/Disciplina/DB/DSL/SimpleTxDB.hs [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+-- | Simple transaction database
+newtype SimpleTxDB a = SimpleTxDB
+   { getSimpleTxDB :: Reader [PrivateTx] a
+   } deriving (Functor, Applicative, Monad, MonadReader [PrivateTx])
+
+instance MonadSearchTxObj SimpleTxDB where
+  runTxQuery = runSimpleTxQuery
+  runTxsQuery = runSimpleTxsQuery
+  runObjQuery = undefined -- TODO, implement this one, but how are Object defined?
```
Comment (flyingleafe) [inline src/Disciplina/DB/DSL/SimpleTxDB.hs]: It's a good question, but for now we can assume that `Object` ~ `LByteString`, I think.

File: tests/Test/Disciplina/DB/DSL/SimpleTxDB.hs [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+import Test.Common
+
+import Crypto.Error (CryptoFailable(..))
+import Disciplina.DB (QueryTx(..), QueryTxs(..), WHERE(..)
+                     ,TxIdEq(..), TxGrade(..)
+                     ,TxsFilterExpr(..), runSimpleTxDBQuery)
+import Disciplina.Educator (PrivateTx(..), EducatorTxMsg(..)
+                           ,StudentTxMsg(..), PrivateTxPayload(..))
+import Disciplina.Crypto (hash, AbstractPK(..), AbstractSK(..), PublicKey, SecretKey)
```
Comment (lambda76) [inline tests/Test/Disciplina/DB/DSL/SimpleTxDB.hs]: I don't really want to import AbstractPK, AbstractSK here but didn't find another way to create PublicKey, SecretKey.

File: tests/Test/Disciplina/DB/DSL/Interpret/SimpleTxDB.hs [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+mkPrivateTx courseId payload studentKey educatorKey =
+  PrivateTx {
+       _ptxStudentId = mkAddr . fst $ mkKeyPair studentKey
+     , _ptxCourseId = courseId
+     , _ptxEducatorId = mkAddr . fst $ mkKeyPair educatorKey
+     , _ptxPayload = payload
+     }
+ where
+   -- Create key pair from seed
+   mkKeyPair :: Char -> (PublicKey, SecretKey)
```
Comment (flyingleafe) [inline tests/Test/Disciplina/DB/DSL/Interpret/SimpleTxDB.hs]: Thinking about it, I understood that there should possibly be a method in `SignatureScheme` class for keypair generation from a seed. But this is out of scope of this PR.
